### ISSUE TITLE
collect options into a single get/set function

### ISSUE THREAD

**Issue by mariev** [phuse-org/valtools#19]
current state:

code in branch `dev_eli`, `dev_eli_ellis_exploration` use a combination of functions to get/set valtools options:

- `vt_username()` - problematic name b/c no verb....; why does "vt_username" have a separate function?
- `getOption("vt.validation_directory",default = "vignettes/validation")`, `getOption("vt.validate_install_directory", default = "validation")` , `getOption("vt.validation_output_directory",default = "inst/validation")` - problematic b/c sets state *within* a function for shared attribute (see render_validation_report.R,  init.R, utils.R)
- initial options set via `.onload` - makes vt options largely invisible to advanced users without manually reviewing `options()`

recommendation:

- adopt a `vt_options` feature similar to knitr `opts_chunk` & `new_defaults` example: https://github.com/yihui/knitr/blob/master/R/defaults.R

**Comment by thebioengineer** [comment]
Another thing we can think about too is user-level and project-level configs:
https://blog.r-hub.io/2020/03/12/user-preferences/

user level configs: name used when creating a document
project level configs: base dir when creating documents

**Comment by thebioengineer** [comment]
how does the following behavior sound:

for setting project level configs:
- create a .validation config file
- add .validation to the .Rbuildignore (depending on decision below)
- the config file contains
    - validation working dir (`vignettes/validation`)
    - validation results dir (`inst/validation`) <-Do we want to allow this to be editable if working on a package? If so, this information needs to be passed into the built R package as well...

for setting user level configs
- use caching so the user can set their username via `vt_set_username()`, and access it via `vt_username`? (similar to  `github_pat()` gets your github pat for you when using `remotes::install_github()`.)
- if no cache is set, it falls back to the computer username, if that fails, ""

My concern with the opts_chunk() is that the library needs to be loaded before it has any impact on the environment, and the user then needs to set them on loading of the package.

**Comment by mariev** [comment]
I am wondering for the user-level configs if we can add a pointer via the .validation config file to the cache/computer username location? Trying to keep things dry but not have to remember multiple places....

**Comment by thebioengineer** [comment]
[Initial look](https://github.com/phuse-org/valtools/blob/validation_config/R/config.R) at setting up a config file with the information about users as well, so it is always a project level setting. 

Creates a .validation file at the base directory that contains a json that has: working validation directory, the output directory (where in inst) to copy when using `vt_validated_build()`, and a username_list, that will contain a key for username, preferred name, and title. This might be helpful for populating tables for the validation report?

**Comment by mariev** [comment]
I think there is something going on in the `sapply()` call from `check_username_list`:
```
> library(valtools)
> 
> vt_use_validation_config()

Please supply some information for recording users within the package.
Note, that this information can be updated at any time though `vt_add_user_to_config()`

 Please provide your name associated with the username `mvendett` and press `Enter`: Marie

 Please provide your title associated with the username `mvendett` and press `Enter`: statprog

Error in check_username_list(username_list) : 
  'list' object cannot be coerced to type 'logical'
In addition: Warning message:
In all(sapply(seq_along(username_list), function(user_info_idx,  :
  coercing argument of type 'list' to logical
> traceback()
2: check_username_list(username_list) at config.R#84
1: vt_use_validation_config()
```

**Comment by thebioengineer** [comment]
pushed an update. should be good now. I also removed the requirement that a user be added on initialization of the config:

```
> library(valtools)
> vt_use_validation_config()
validation config file created. Add user information through `vt_add_user_to_config()`
> vt_add_user_to_config()

Please supply some information for recording users within the package.
Note, that this information can be updated at any time though `vt_add_user_to_config()`

 Please provide your name associated with the username `ehhughes` and press `Enter`: Ellis Hughes

 Please provide your title associated with the username `ehhughes` and press `Enter`: Statistical Programmer

User `ehhughes` added to validation config file.
> readLines(".validation")
 [1] "{"                                                      
 [2] "  \"validation_directory\": [\"vignettes/validation\"],"
 [3] "  \"validation_output_directory\": [\"validation\"],"   
 [4] "  \"usernames\": {"                                     
 [5] "    \"ehhughes\": {"                                    
 [6] "      \"name\": [\"Ellis Hughes\"],"                    
 [7] "      \"title\": [\"Statistical Programmer\"],"         
 [8] "      \"username\": [\"ehhughes\"]"                     
 [9] "    }"                                                  
[10] "  }"                                                    
[11] "}"   
```

**Comment by thebioengineer** [comment]
we can also add a user without prompts if all the information is passed: 
```
> vt_add_user_to_config(username = "user1", name = "Gary Host", title = "the ghost")
User `user1` added to validation config file.
```

**Comment by mariev** [comment]
🤔 wondering if we want to tweak the structure of .validation a bit to facilitate direct entry? not everyone will want to use the interface

.validation example, current:
```
{
  "validation_directory": ["vignettes/validation"],
  "validation_output_directory": ["validation"],
  "usernames": {
    "mvendett": {
      "name": ["something"],
      "title": ["a person"],
      "username": ["mvendett"]
    },
    "mvendett2": {
      "name": ["myname"],
      "title": ["mytitle"],
      "username": ["mvendett2"]
    }
  }
}
```

.validation example, simplified - striving for a yaml-style syntax
```
validation_directory:
  vignettes/validation
validation_output_directory:
  validation
usernames:
  user1:
    name: My Name
    title: My title
  user2:
    name: Second Name
    title: A title
```

**Comment by thebioengineer** [comment]
I think that makes sense! I've updated the code and tests to use yaml instead.
Would `validation_output_directory` be better as `validation_inst_directory`? It really will only impact cases where the user builds the package.

**Comment by thebioengineer** [comment]
@elimillera - I figure this will be added into the `vt_create_package()` function flow.

**Comment by mariev** [comment]
@thebioengineer I like the switch to yaml!  

- Wondering why we need username in two places?
- What happens if .validation exists and I call `vt_use_validation_config`? Do I quietly lose existing .validation entries?
- Should there be an early exit if I am trying to run `vt_add_user_to_config()` without .validation present? 🤔 maybe the error should trigger `vt_use_validation_config()`?
- is .validation added to .Rbuildignore?
- super nitpicky: why is the verb "add" for both adding new users & updating?
- scope creep: can we use the validation_directory, validation_output_directory values to identify when we are using the different validation models? e.g. to vignettes vs vignettes+inst?

**Comment by thebioengineer** [comment]
> Wondering why we need username in two places?

I was thinking in the case that things are manually entered I wanted to have a double check I guess. 

> What happens if .validation exists and I call vt_use_validation_config? Do I quietly lose existing .validation entries?

currently, yes. working on a fix for that

> Should there be an early exit if I am trying to run vt_add_user_to_config() without .validation present? 🤔 maybe the error should trigger vt_use_validation_config()?

working on a fix. in my working copy I have this error message returned: 
```
Error: A validation config file does not exist.
Run `valtools::vt_use_validation_config()` to create a validation config file.
```
 
> is .validation added to .Rbuildignore?

not right now, though on testing building it did not get kept, so I have a feeling it would be dropped? but we can make it explicitly added. 

> super nitpicky: why is the verb "add" for both adding new users & updating?

no particular reason other than being lazy. figured less functions to remember?

> scope creep: can we use the validation_directory, validation_output_directory values to identify when we are using the different validation models? e.g. to vignettes vs vignettes+inst?

Kind of. Mentally, I have been kind of approaching this from the idea that we want to get people to follow the "work in vignettes" process, but automagically supports the "vignettes+inst" process if they want to do validation on installed packages.

I will ping you when I have an update pushed re: overwriting

**Comment by thebioengineer** [comment]
@mariev, check out 465696b

I also added `role` as a field to capture. so we in theory would have all the info necessary around all the users for the project to construct the tables for signature

**Comment by mariev** [comment]
so I think I am eyeballing this .validation config file with a slightly different intent than original accessor functions this issue was mentioning, one which I believe is worth pursuing. For now I am going to focus on the `usernames:` slot, which I am kind of seeing as a cross between DESCRIPTION Authors and bibliography entries.

1. What makes up a record?
    - Display name
    - role (1+)
    - username 
    - handle (shortname to reference a record), currently matches username

2. What records are included?
    - each person who will sign off on validation report
    - people who were involved in validation for prior releases?
    - people who are involved in validation generally across an org?

3. How to use username record detail?
    - verify that validation lead is the R username for current session. 🤔 wondering if we need to accommodate for something different, eg. if validation is run as part of c.i.?
    - cross referencing between signatures and names that come up for function/requirement/test case/test code authorship
---

I think the latest code is getting close, but `vt_add_user_to_config()` needs a bit more polish. Specifically, updating/adding role? more than one role?

---
I still don't get why it is important that usage is: 
```
vt_use_validation_config(username_list = list(mvendett = list(username = "mvendett", 
                                                                  name = "Marie", 
                                                                  role = "validation lead", 
                                                                  title = "statprog")), 
                           overwrite = TRUE)
```

why not

```
vt_use_validation_config(username_list = list(username = "mvendett", 
                                                  name = "Marie", 
                                                  role = "validation lead", 
                                                  title = "statprog"), 
                           overwrite = TRUE)

``` 

with .username records as

```
usernames:
  mvendett:
    name: Marie
    role: validation lead
    title: statprog
```

**Comment by mariev** [comment]
also wanted to note that if .validation is in the pkg directory it will give a NOTE at cran check, identified as a hidden file.

**Comment by thebioengineer** [comment]
based on our conversation:
- simplify the username_list to example above
- change `validation_directory` to `working_directory` <- reflects where "work" gets done
- change `validation_output_directory` to `output_directory` <- reflects if content should be copied to a different location on build
- set `working_directory` to just "vignettes", user will not be able to decide if the folders name is "validation". allows user to decide if they want to just work in "inst", or rando network drive, etc.
- set `output_directory` to "inst". if "" or FALSE, then all content stays in working dir
- add `report_naming_format` option, allows users to set the output naming format w/o extension of the validation report. 
- set `report_naming_format` to "Validation_Report_{package}_v{version}_{date}"
- see about scraping DESCRIPTION for Authors@R to populate config from existing projects

**Comment by thebioengineer** [comment]
Addressed everything but DESCRIPTION in the latest push.

**Comment by thebioengineer** [comment]
This was added with #26. However, we now need to go back and add this ability into the package.

blocked by #32

**Comment by mariev** [comment]
@thebioengineer we're talking about adding this to examples/tests, yes? or something else?

**Comment by thebioengineer** [comment]
I meant going through the package and updating all the `getOption()`'s throughout the code

**Comment by mariev** [comment]
closed via #34
